File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/map/CustomGeometrySourceState.java
package org.maplibre.nativeffi.map;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import org.maplibre.nativeffi.error.InvalidArgumentException;
import org.maplibre.nativeffi.geo.CanonicalTileId;
import org.maplibre.nativeffi.internal.c.MapLibreNativeC;
import org.maplibre.nativeffi.internal.c.mln_canonical_tile_id;
import org.maplibre.nativeffi.internal.c.mln_custom_geometry_source_options;
import org.maplibre.nativeffi.internal.c.mln_custom_geometry_source_tile_callback;
import org.maplibre.nativeffi.internal.callback.CallbackLifecycle;
import org.maplibre.nativeffi.style.CustomGeometrySourceOptions;

/** Owns map/style-scoped custom geometry source callback state. */
final class CustomGeometrySourceState implements AutoCloseable {
  private final Arena arena;
  private final CustomGeometrySourceOptions options;
  private final MemorySegment fetchStub;
  private final MemorySegment cancelStub;
  private final MemorySegment descriptor;
  private final CallbackLifecycle lifecycle = new CallbackLifecycle();

  CustomGeometrySourceState(CustomGeometrySourceOptions options) {
    this.arena = Arena.ofShared();
    this.options = options;
    this.fetchStub = mln_custom_geometry_source_tile_callback.allocate(this::fetchTile, arena);
    this.cancelStub = mln_custom_geometry_source_tile_callback.allocate(this::cancelTile, arena);
    this.descriptor = MapLibreNativeC.mln_custom_geometry_source_options_default(arena);
    mln_custom_geometry_source_options.fetch_tile(descriptor, fetchStub);
    mln_custom_geometry_source_options.cancel_tile(descriptor, cancelStub);
    mln_custom_geometry_source_options.user_data(descriptor, MemorySegment.NULL);
    writeFields();
  }

  MemorySegment descriptor() {
    return descriptor;
  }

  private void writeFields() {
    if ((options.hasTileSize() && options.tileSize() < 0)
        || (options.hasBuffer() && options.buffer() < 0)) {
      throw new InvalidArgumentException(
          MapLibreNativeC.MLN_STATUS_INVALID_ARGUMENT(),
          "custom geometry source unsigned options must be non-negative");
    }
    var fields = 0;
    if (options.hasMinZoom()) {
      fields |= MapLibreNativeC.MLN_CUSTOM_GEOMETRY_SOURCE_OPTION_MIN_ZOOM();
      mln_custom_geometry_source_options.min_zoom(descriptor, options.minZoom());
    }
    if (options.hasMaxZoom()) {
      fields |= MapLibreNativeC.MLN_CUSTOM_GEOMETRY_SOURCE_OPTION_MAX_ZOOM();
      mln_custom_geometry_source_options.max_zoom(descriptor, options.maxZoom());
    }
    if (options.hasTolerance()) {
      fields |= MapLibreNativeC.MLN_CUSTOM_GEOMETRY_SOURCE_OPTION_TOLERANCE();
      mln_custom_geometry_source_options.tolerance(descriptor, options.tolerance());
    }
    if (options.hasTileSize()) {
      fields |= MapLibreNativeC.MLN_CUSTOM_GEOMETRY_SOURCE_OPTION_TILE_SIZE();
      mln_custom_geometry_source_options.tile_size(descriptor, options.tileSize());
    }
    if (options.hasBuffer()) {
      fields |= MapLibreNativeC.MLN_CUSTOM_GEOMETRY_SOURCE_OPTION_BUFFER();
      mln_custom_geometry_source_options.buffer(descriptor, options.buffer());
    }
    if (options.hasClip()) {
      fields |= MapLibreNativeC.MLN_CUSTOM_GEOMETRY_SOURCE_OPTION_CLIP();
      mln_custom_geometry_source_options.clip(descriptor, options.clip());
    }
    if (options.hasWrap()) {
      fields |= MapLibreNativeC.MLN_CUSTOM_GEOMETRY_SOURCE_OPTION_WRAP();
      mln_custom_geometry_source_options.wrap(descriptor, options.wrap());
    }
    mln_custom_geometry_source_options.fields(descriptor, fields);
  }

  private void fetchTile(MemorySegment userData, MemorySegment tileId) {
    var lease = lifecycle.enter();
    if (lease.isEmpty()) {
      return;
    }
    try (var ignored = lease.get()) {
      options.callback().fetchTile(canonicalTileId(tileId));
    } catch (Throwable ignored) {
      // Native callbacks must not unwind through the C ABI.
    }
  }

  private void cancelTile(MemorySegment userData, MemorySegment tileId) {
    var lease = lifecycle.enter();
    if (lease.isEmpty()) {
      return;
    }
    try (var ignored = lease.get()) {
      options.callback().cancelTile(canonicalTileId(tileId));
    } catch (Throwable ignored) {
      // Native callbacks must not unwind through the C ABI.
    }
  }

  private static CanonicalTileId canonicalTileId(MemorySegment segment) {
    return new CanonicalTileId(
        mln_canonical_tile_id.z(segment),
        Integer.toUnsignedLong(mln_canonical_tile_id.x(segment)),
        Integer.toUnsignedLong(mln_canonical_tile_id.y(segment)));
  }

  @Override
  public void close() {
    lifecycle.close("Custom geometry source", arena::close);
  }
}

File: bindings/kotlin/src/nativeMain/kotlin/org/maplibre/nativeffi/map/CustomGeometrySourceState.kt
package org.maplibre.nativeffi.map

import kotlinx.cinterop.COpaquePointer
import kotlinx.cinterop.CPointer
import kotlinx.cinterop.CValue
import kotlinx.cinterop.ExperimentalForeignApi
import kotlinx.cinterop.StableRef
import kotlinx.cinterop.alloc
import kotlinx.cinterop.asStableRef
import kotlinx.cinterop.nativeHeap
import kotlinx.cinterop.ptr
import kotlinx.cinterop.staticCFunction
import kotlinx.cinterop.useContents
import org.maplibre.nativeffi.internal.c.MLN_CUSTOM_GEOMETRY_SOURCE_OPTION_BUFFER
import org.maplibre.nativeffi.internal.c.MLN_CUSTOM_GEOMETRY_SOURCE_OPTION_CLIP
import org.maplibre.nativeffi.internal.c.MLN_CUSTOM_GEOMETRY_SOURCE_OPTION_MAX_ZOOM
import org.maplibre.nativeffi.internal.c.MLN_CUSTOM_GEOMETRY_SOURCE_OPTION_MIN_ZOOM
import org.maplibre.nativeffi.internal.c.MLN_CUSTOM_GEOMETRY_SOURCE_OPTION_TILE_SIZE
import org.maplibre.nativeffi.internal.c.MLN_CUSTOM_GEOMETRY_SOURCE_OPTION_TOLERANCE
import org.maplibre.nativeffi.internal.c.MLN_CUSTOM_GEOMETRY_SOURCE_OPTION_WRAP
import org.maplibre.nativeffi.internal.c.mln_canonical_tile_id
import org.maplibre.nativeffi.internal.c.mln_custom_geometry_source_options
import org.maplibre.nativeffi.internal.c.mln_custom_geometry_source_options_default
import org.maplibre.nativeffi.internal.callback.CallbackGate
import org.maplibre.nativeffi.internal.struct.StyleStructs
import org.maplibre.nativeffi.style.CustomGeometrySourceOptions

/** Owns map/style-scoped custom geometry source callback state. */
@OptIn(ExperimentalForeignApi::class)
internal class CustomGeometrySourceState(private val options: CustomGeometrySourceOptions) :
  AutoCloseable {
  private val selfRef = StableRef.create(this)
  private val descriptor = nativeHeap.alloc<mln_custom_geometry_source_options>()
  private val gate = CallbackGate("custom geometry callbacks") { closeNative() }

  init {
    mln_custom_geometry_source_options_default().place(descriptor.ptr)
    descriptor.fetch_tile = staticCFunction(::customGeometryFetchTile)
    descriptor.cancel_tile = staticCFunction(::customGeometryCancelTile)
    descriptor.user_data = selfRef.asCPointer()
    writeFields(descriptor)
  }

  fun descriptor(): CPointer<mln_custom_geometry_source_options> = descriptor.ptr

  internal fun fetch(tileId: CValue<mln_canonical_tile_id>) {
    val lease = gate.enter() ?: return
    try {
      options.callback.fetchTile(tileId.useContents { StyleStructs.canonicalTileId(this) })
    } catch (_: Throwable) {
      // Native callbacks must not unwind through the C ABI.
    } finally {
      lease.close()
    }
  }

  internal fun cancel(tileId: CValue<mln_canonical_tile_id>) {
    val lease = gate.enter() ?: return
    try {
      options.callback.cancelTile(tileId.useContents { StyleStructs.canonicalTileId(this) })
    } catch (_: Throwable) {
      // Native callbacks must not unwind through the C ABI.
    } finally {
      lease.close()
    }
  }

  private fun writeFields(native: mln_custom_geometry_source_options) {
    native.fields = 0U
    options.minZoom?.let {
      native.fields = native.fields or MLN_CUSTOM_GEOMETRY_SOURCE_OPTION_MIN_ZOOM
      native.min_zoom = it
    }
    options.maxZoom?.let {
      native.fields = native.fields or MLN_CUSTOM_GEOMETRY_SOURCE_OPTION_MAX_ZOOM
      native.max_zoom = it
    }
    options.tolerance?.let {
      native.fields = native.fields or MLN_CUSTOM_GEOMETRY_SOURCE_OPTION_TOLERANCE
      native.tolerance = it
    }
    options.tileSize?.let {
      native.fields = native.fields or MLN_CUSTOM_GEOMETRY_SOURCE_OPTION_TILE_SIZE
      native.tile_size = it.toUInt()
    }
    options.buffer?.let {
      native.fields = native.fields or MLN_CUSTOM_GEOMETRY_SOURCE_OPTION_BUFFER
      native.buffer = it.toUInt()
    }
    options.clip?.let {
      native.fields = native.fields or MLN_CUSTOM_GEOMETRY_SOURCE_OPTION_CLIP
      native.clip = it
    }
    options.wrap?.let {
      native.fields = native.fields or MLN_CUSTOM_GEOMETRY_SOURCE_OPTION_WRAP
      native.wrap = it
    }
  }

  override fun close() = gate.close()

  private fun closeNative() {
    selfRef.dispose()
    nativeHeap.free(descriptor.rawPtr)
  }
}

@Suppress("UNUSED_PARAMETER")
@OptIn(ExperimentalForeignApi::class)
private fun customGeometryFetchTile(
  userData: COpaquePointer?,
  tileId: CValue<mln_canonical_tile_id>,
) {
  userData?.asStableRef<CustomGeometrySourceState>()?.get()?.fetch(tileId)
}

@Suppress("UNUSED_PARAMETER")
@OptIn(ExperimentalForeignApi::class)
private fun customGeometryCancelTile(
  userData: COpaquePointer?,
  tileId: CValue<mln_canonical_tile_id>,
) {
  userData?.asStableRef<CustomGeometrySourceState>()?.get()?.cancel(tileId)
}
